hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/fsck/ContainerHealthTask.java

```diff
@@ -0,0 +1,345 @@
+    try {
+      if (processedContainers.contains(container)) {
+        // Was already handled when processing the existing DB records, so skip
+        return;
+      }
+      Set<ContainerReplica> containerReplicas =
+          containerManager.getContainerReplicas(container.containerID());
+      ContainerHealthStatus h = new ContainerHealthStatus(
+          container, containerReplicas, placementPolicy);
+      if (h.isHealthy()) {
+        return;
+      }
+      containerSchemaManager.insertUnhealthyContainerRecords(
+          ContainerHealthRecords.generateUnhealthyRecords(h, currentTime));
+    } catch (ContainerNotFoundException e) {
+      LOG.error("Container not found while finding missing containers", e);
+    }
+  }
+
+  /**
+   * Helper methods to generate and update the required database records for
+   * unhealthy containers.
+   */
+  static public class ContainerHealthRecords {
+
+    static public boolean retainOrUpdateRecord(
+        ContainerHealthStatus container, UnhealthyContainersRecord rec) {
+      boolean returnValue = false;
+      switch(UnHealthyContainerStates.valueOf(rec.getContainerState())) {
+      case MISSING:
+        returnValue = container.isMissing();
+        break;
+      case MIS_REPLICATED:
+        returnValue = keepMisReplicatedRecord(container, rec);
+        break;
+      case UNDER_REPLICATED:
+        returnValue = keepUnderReplicatedRecord(container, rec);
+        break;
+      case OVER_REPLICATED:
+        returnValue = keepOverReplicatedRecord(container, rec);
+        break;
+      default:
+        returnValue = false;
+      }
+      return returnValue;
+    }
+
+    static public List<UnhealthyContainers> generateUnhealthyRecords(
+        ContainerHealthStatus container, long time) {
+      return generateUnhealthyRecords(container, new HashSet<>(), time);
+    }
+
+    /**
+     * Check the status of the container and generate any database records that
+     * need to be recorded. This method also considers the records seen by the
+     * method retainOrUpdateRecord. If a record has been seen by that method
+     * then it will not be emitted here. Therefore this method returns only the
+     * missing records which have not been seen already.
+     * @return List of UnhealthyContainer records to be stored in the DB
+     */
+    static public List<UnhealthyContainers> generateUnhealthyRecords(
+        ContainerHealthStatus container, Set<String> recordForStateExists,
+        long time) {
+      List<UnhealthyContainers> records = new ArrayList<>();
+      if (container.isHealthy()) {
+        return records;
+      }
+
+      if (container.isMissing()
+          && !recordForStateExists.contains(
+              UnHealthyContainerStates.MISSING.toString())) {
+        records.add(
+            recordForState(container, UnHealthyContainerStates.MISSING, time));
+      }
+
+      if (container.isUnderReplicated()
```

Comment:
Can we add an 'else' after the isMissing() check? A missing container cannot be in any other state right?

Reply:
You are correct. Rather than next the rest of the method in the else block, I added a return statement in the missing branch. Does that look ok?

Reply:
+1